src/core/derivation.ts

```diff
@@ -155,12 +160,23 @@ export function checkIfStateModificationsAreAllowed(atom: IAtom) {
         )
 }
 
+export function checkIfStateReadsAreAllowed(observable: IObservable) {
+    if (
+        process.env.NODE_ENV !== "production" &&
+        !globalState.allowStateReads &&
+        globalState.observableRequiresReaction
+    ) {
+        console.warn(`[mobx] Observable ${observable.name} being read outside a reactive context`)
```

Comment:
Maybe make the error a bit clearer for those not fully familiar with these concept names? ".. is being read outside an action or reactive context, such as 'autorun', 'reaction', 'when' or 'observer'"

Reply:
I want to point out that when using react for example, that message can be confusing.
The developer might thow an `autorun` inside render to make it go away

BTW I took that message from:
https://github.com/mobxjs/mobx/blob/master/src/core/computedvalue.ts#L282-L287
So maybe we should make the change in both place

Reply:
Agreed, should probably change it in both places. Also, in this case, we might also want to say something along the lines: "... or mark the surrounding function as `action` to indicate that there are imperative state reads and writes"? Not sure how to phrase it right, (maybe we should link to some document to have more explanation room), but I am concerned about the case discussed here: https://github.com/mobxjs/mobx/pull/2079#discussion_r328116458

If something is intended to be an action, but not marked, then getting the above error on the first read that happens in the 'action' might be very misleading.

Reply:
I'm also not very good with message :)
these options are opt-it,
Maybe we should just tell the developer to look for `observableRequiresReaction` and elaborate in the docs.

>observableRequiresReaction violation: observable ${observable.name} being read outside a reactive context

Reply:
As these feature marked as experimental, i would prefer to push it forward and get feedback, and then improve developer experience